### PR TITLE
Rewrites/dba array string table

### PR DESCRIPTION
**Pull request by Quix0r**
Rewrite:
- avoid having `array|string` for `$table` to have a strong `string` type-hint for `$table`
- you now have to do it for yourself by giving `schema.table` as parameter

### CODE REVIEW

**Comment by Quix0r** [comment]
**SOLVED**: I have this here: `Table 'test.information_schema.tables' doesn't exist `

**Comment by Quix0r** [comment]
**SOLVED**: Also strange is that `DROP VIEW IF EXISTS` is being removed from `database.sql`?

**Comment by Quix0r** [comment]
**SOLVED**: In `Database/View::createView()` a call to `self::isView()` might have prevented `DROP VIEW IF EXISTS` being added.

**Comment by Quix0r** [comment]
And can we move `isView/Table()` to `Database` class instead? They feel very generic to me.

**Comment by Quix0r** [comment]
Okay, I have added a debug logger for the SQL statement being used and I got `NFORMATION_SCHEMA.TABLES` being wrapped into back-ticks.  That needs fixing!

**Comment by Quix0r** [comment]
Okay, I have added a debug logger for the SQL statement being used and I got `NFORMATION_SCHEMA.TABLES` being wrapped into back-ticks.  That needs fixing!

File: src/Database/DBA.php [hunk 1]
```diff
@@ -524,20 +521,20 @@ public static function count($table, array $condition = [], array $params = []):
...
-		if (is_string($tables)) {
-			$tables = [$tables];
-		}
-
 		$quotedTables = [];
 
 		foreach ($tables as $schema => $table) {
 			if (is_numeric($schema)) {
-				$quotedTables[] = self::quoteIdentifier($table);
+				$str = '';
```
Comment (Quix0r) [inline src/Database/DBA.php]: Well, some quick'n'dirty code here. Normally only 1 or 2 elements should be expected here.

**Comment by Quix0r** [comment]
Now getting `Error: Call to a member function rollBack() on null` when local database configuration is not valid. This originates from `DatabaseTestTrait::setUpDb()`. It looks like the code there assumes that the database connection is established.

File: tests/Util/Database/StaticDatabase.php [hunk 2]
```diff
@@ -179,7 +180,8 @@ public static function statConnect(array $server)
 			self::$staticConnection = @new ExtendedPDO($connect, $user, $pass);
 			self::$staticConnection->setAttribute(PDO::ATTR_AUTOCOMMIT,0);
 		} catch (PDOException $e) {
-			/// @TODO At least log exception, don't ignore it!
+			// Log exception
+			Logger::error('Cannot establish database connection', ['exception' => $e, 'host' => $server, 'dbname' => $db]);
```
Comment (nupplaphil) [inline tests/Util/Database/StaticDatabase.php]: Plz no dependency injection directly without using the constructor, because that's where the party gets started :)

`Logger::notice()` is calling `self::logger->notice()`, which has to be initialized with proper `LoggerInterface $logger` which is created by a factory using... the `Database` itself :)) and `StaticDatabase` must be called before any DB call, otherwise there may be a transaction mismatching during tests

**Review by nupplaphil — CHANGES_REQUESTED** [review]

File: bin/dev/run_tests.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
```
Comment (nupplaphil) [inline bin/dev/run_tests.sh]: Did you give `autotest.sh` in the root path a try? :)

**Review by nupplaphil — CHANGES_REQUESTED** [review]

File: bin/dev/run_tests.sh (hunk 3 above)
Comment (Quix0r) [reply bin/dev/run_tests.sh]: The script `autotest.sh` requires to set environment variables `MYSQL_*` which are okay for a single project development (e.g. being set in `~/.bashrc`) but not for multi-project developments. This is why I had prefixed my variables with `FRIENDICA_` so they are Friendica-specific.

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: So I better use `$this->logger` instead?

**Comment by MrPetovan** [comment]
> * avoid having `array|string` for `$table` to have a strong `string` type-hint for `$table`

I disagree with the "strong". `string` is the widest type-hint you can have for a parameter and as such, `'(schema.)table'` is actually less specific than `'table'|['schema' => 'table']`. 

You can see the issue with `DBA::buildTableString` where you have to perform several awkward strings operations (`explode()`, `rtrim()`) that weren't necessary with the previous data structure.

You want a strong type-hint, create a `DatabaseTableName` class that will have an optional schema property, then you will have a strong type-hint.

If you want to stick with strings, I suggest you do the string manipulation in `DBA::quoteIdentifier` where you can split on periods and concatenate the result better than you're currently doing in `DBA::buildTableString`:
```php
public static function quoteIdentifier(string $identifier): string
{
  return implode(
    '.',
    array_map(
      function (string $identifier) { return '`' . str_replace('`', '``', $identifier) . '`'; },
      explode('.', $identifier)
    )
  );
}
```
And this way `DBA::buildTableString` becomes just:
```php
public static function buildTableString(array $tables): string
{
  return implode(',', array_map(['self', 'quoteIdentifier'], $tables));
}
```
Either way, please don't leave the codebase in the current state of your branch, it's the worst of both world.

**Comment by Quix0r** [comment]
With the old version `DBA::buildTableString()` would have no type-hint for the parameter, allowing all variables, including boolean and objects. `string` at least only allows strings.

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: No, there is no object to use `$this`. :-( Ops, my mistake.

Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: So, how do I get a logger in a static (non-object) method? `Logger::foo()` seem to have worked. At least I got the message logged.

Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: If we could make `Database::$logger` static it would solve this? Plus rewrite all invocations to `self::$logger` of course.

File: src/Database/DBA.php [hunk 4]
```diff
@@ -218,12 +218,12 @@ public static function e(string $sql): bool
 	/**
 	 * Check if data exists
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array               $condition array of fields for condition
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 5]
```diff
@@ -305,13 +304,12 @@ public static function insert($table, array $param, int $duplicate_mode = Databa
 	 * Inserts a row with the provided data in the provided table.
 	 * If the data corresponds to an existing row through a UNIQUE or PRIMARY index constraints, it updates the row instead.
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array        $param parameter array
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 6]
```diff
@@ -331,12 +329,11 @@ public static function lastInsertId(): int
 	 *
 	 * This function can be extended in the future to accept a table array as well.
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 7]
```diff
@@ -385,13 +382,13 @@ public static function rollback(): bool
 	/**
 	 * Delete a row from a table
 	 *
-	 * @param string|array $table      Table name
-	 * @param array        $conditions Field condition(s)
+	 * @param string $table      Table name
```
Comment (MrPetovan) [inline src/Database/DBA.php:385]: Please update the documentation to reflect the type change (the possible schema prefix).

File: src/Database/DBA.php [hunk 8]
```diff
@@ -417,7 +414,7 @@ public static function delete($table, array $conditions, array $options = []): b
 	 * Only set $old_fields to a boolean value when you are sure that you will update a single row.
 	 * When you set $old_fields to "true" then $fields must contain all relevant fields!
 	 *
-	 * @param string|array  $table      Table name or array [schema => table]
+	 * @param string        $table      Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 9]
```diff
@@ -450,27 +447,27 @@ public static function selectFirst($table, array $fields = [], array $condition
 	/**
 	 * Select rows from a table and fills an array with the data
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 10]
```diff
@@ -450,27 +447,27 @@ public static function selectFirst($table, array $fields = [], array $condition
...
 	}
 
 	/**
 	 * Select rows from a table
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 11]
```diff
@@ -487,17 +484,17 @@ public static function selectToArray($table, array $fields = [], array $conditio
...
 		return DI::dba()->select($table, $fields, $condition, $params);
 	}
 
 	/**
 	 * Counts the rows from a table satisfying the provided condition
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Please update the documentation to reflect the type change.

File: src/Database/DBA.php [hunk 12]
```diff
@@ -524,37 +521,30 @@ public static function count($table, array $condition = [], array $params = []):
 	 * - [table1, table2, ...]
 	 * - [schema1 => table1, schema2 => table2, table3, ...]
```
Comment (MrPetovan) [inline src/Database/DBA.php:522]: Please update the documentation to reflect the type change.

File: src/Database/DBStructure.php [hunk 13]
```diff
@@ -1064,7 +1064,7 @@ function ($to, $from) {
 	 * @return boolean Does the table exist?
 	 * @throws Exception
 	 */
-	public static function existsColumn($table, $columns = [])
+	public static function existsColumn(string $table, array $columns = []): bool
```
Comment (MrPetovan) [inline src/Database/DBStructure.php:1067]: Please update the documentation to reflect the type change with the optional schema prefix.

File: src/Database/Database.php [hunk 14]
```diff
@@ -64,7 +64,7 @@ class Database
 	protected $server_info    = '';
 	/** @var PDO|mysqli */
 	protected $connection;
-	protected $driver;
+	protected $driver = '';
```
Comment (MrPetovan) [inline src/Database/Database.php:67]: Why the change to empty string?

File: src/Database/Database.php [hunk 15]
```diff
@@ -820,13 +820,13 @@ public function e(string $sql): bool
 	/**
 	 * Check if data exists
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition array of fields for condition
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 16]
```diff
@@ -1003,14 +1003,14 @@ public function fetch($stmt)
 	/**
 	 * Insert a row into a table. Field value objects will be cast as string.
 	 *
-	 * @param string|array $table          Table name or array [schema => table]
-	 * @param array        $param          parameter array
-	 * @param int          $duplicate_mode What to do on a duplicated entry
+	 * @param string $table          Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 17]
```diff
@@ -1054,13 +1054,12 @@ public function insert($table, array $param, int $duplicate_mode = self::INSERT_
 	 * Inserts a row with the provided data in the provided table.
 	 * If the data corresponds to an existing row through a UNIQUE or PRIMARY index constraints, it updates the row instead.
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array        $param parameter array
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 18]
```diff
@@ -1103,12 +1102,11 @@ public function lastInsertId(): int
 	 *
 	 * This function can be extended in the future to accept a table array as well.
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: What I have also noticed in that method is that a lot variables are **only** initialized when the first `if()` block is executed. If it is not executed they (e.g. `$db_host`, `$db_user`, `$db_pw`) are not initialized but used later leading maybe to "Undefined variable" errors. If we could initialize all our variables clearly, means on the method start, this can be prevented.

File: src/Database/Database.php [hunk 19]
```diff
@@ -1265,14 +1263,14 @@ public function rollback(): bool
 	 * @return boolean was the delete successful?
 	 * @throws \Exception
 	 */
-	public function delete($table, array $conditions): bool
+	public function delete(string $table, array $conditions): bool
```
Comment (MrPetovan) [inline src/Database/Database.php:1266]: Please update the documentation to reflect the type change with the optional schema prefix.

File: src/Database/Database.php [hunk 20]
```diff
@@ -1302,7 +1300,7 @@ public function delete($table, array $conditions): bool
 	 * Only set $old_fields to a boolean value when you are sure that you will update a single row.
 	 * When you set $old_fields to "true" then $fields must contain all relevant fields!
 	 *
-	 * @param string|array  $table      Table name or array [schema => table]
+	 * @param string        $table      Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 21]
```diff
@@ -1368,16 +1366,16 @@ public function update($table, array $fields, array $condition, $old_fields = []
 	/**
 	 * Retrieve a single record from a table and returns it in an associative array
 	 *
-	 * @param string|array $table
-	 * @param array        $fields
-	 * @param array        $condition
-	 * @param array        $params
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 22]
```diff
@@ -1394,16 +1392,15 @@ public function selectFirst($table, array $fields = [], array $condition = [], a
 	/**
 	 * Select rows from a table and fills an array with the data
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
-	 *
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 23]
```diff
@@ -1469,14 +1466,14 @@ private function escapeFields(array $fields, array $options): array
 	 *
 	 * $data = DBA::select($table, $fields, $condition, $params);
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 24]
```diff
@@ -1509,9 +1506,9 @@ public function select($table, array $fields = [], array $condition = [], array
 	/**
 	 * Counts the rows from a table satisfying the provided condition
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Comment (MrPetovan) [inline src/Database/Database.php]: Please update the documentation to reflect the type change.

File: src/Database/Database.php [hunk 25]
```diff
@@ -1823,14 +1820,14 @@ public function escapeArray(&$arr, bool $add_quotation = false)
 	/**
 	 * Replaces a string in the provided fields of the provided table
 	 *
-	 * @param string $table_name Table name
+	 * @param string $table  Table name
```
Comment (MrPetovan) [inline src/Database/Database.php:1823]: Please update the documentation to reflect the type change with the optional schema prefix.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (nupplaphil) [reply tests/Util/Database/StaticDatabase.php]: There is no easy possibility to add a logger here, that's why there isn't any yet and instead a placeholder.. This execution point is a critical state during a testrun, and tbh I'd like to leave here no further logic (yet) because I spent hours debugging cases, where transactions weren't fully closed and strange/unpredictable errors occur (sometimes -mainly during debugging other errors :) ...)

File: bin/dev/run_tests.sh (hunk 3 above)
Comment (nupplaphil) [reply bin/dev/run_tests.sh]: maybe you cold extend `autorun.sh` beeing more multi-project friendly :-)

tbh I'm not using it very often anymore, so I guess you could rewrite it completely too or drop and replace it ..

File: src/Database/Database.php (hunk 14 above)
Comment (Quix0r) [reply src/Database/Database.php:67]: I have found a few occurrences:
- the getter `getDriver()` states `string` as return type or else `string` cannot be the returned type
- `self::assertContains()` asserts it against strings
- also all `switch()` usages are using only `string` cases.

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (Quix0r) [reply tests/Util/Database/StaticDatabase.php]: Okay, so you suggest me to revert my changes here?

File: bin/dev/run_tests.sh (hunk 3 above)
Comment (Quix0r) [reply bin/dev/run_tests.sh]: Also I would like to move it out of the root directory to `bin/dev/` as this is clearly a developer's script.

File: tests/Util/Database/StaticDatabase.php (hunk 2 above)
Comment (MrPetovan) [reply tests/Util/Database/StaticDatabase.php]: Yes please!

File: tests/Util/Database/StaticDatabase.php [hunk 26]
```diff
@@ -179,8 +179,7 @@ public static function statConnect(array $server)
 			self::$staticConnection = @new ExtendedPDO($connect, $user, $pass);
 			self::$staticConnection->setAttribute(PDO::ATTR_AUTOCOMMIT,0);
 		} catch (PDOException $e) {
-			// Log exception
-			$this->logger::error('Cannot establish database connection', ['exception' => $e, 'host' => $server, 'dbname' => $db]);
+			/// @TODO Try to find a way to log this exception as it contains valueable information
```
Comment (MrPetovan) [inline tests/Util/Database/StaticDatabase.php]: Please add Nupplaphil's comment in the code for the next person who will try to refactor this.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by nupplaphil — APPROVED** [review]

**Review by MrPetovan — APPROVED** [review]
